src/node_report.h

```diff
@@ -0,0 +1,166 @@
+#define PATHSEP "\\"
+#else  // UNIX, OSX
+typedef struct tm TIME_TYPE;
+typedef pid_t PID_TYPE;
+#define PATHSEP "/"
+#endif
+
+void InitializeNodeReport(v8::Isolate* isolate, node::Environment* env);
+
+// Function declarations - functions in src/node_report.cc
+void TriggerNodeReport(v8::Isolate* isolate,
+                       node::Environment* env,
+                       const char* message,
+                       const char* location,
+                       std::string name,
+                       v8::Local<v8::String> stackstr);
+void GetNodeReport(v8::Isolate* isolate,
+                   node::Environment* env,
+                   const char* message,
+                   const char* location,
+                   v8::Local<v8::String> stackstr,
+                   std::ostream& out);
+
+// Function declarations - utility functions in src/utilities.cc
+void ReportEndpoints(uv_handle_t* h, std::ostringstream& out);
+void WalkHandle(uv_handle_t* h, void* arg);
+std::string EscapeJsonChars(const std::string& str);
+
+// Function declarations - export functions in src/node_report_module.cc
+void TriggerReport(const v8::FunctionCallbackInfo<v8::Value>& info);
+void GetReport(const v8::FunctionCallbackInfo<v8::Value>& info);
+
+
+// Node.js boot time - defined in src/node_report.cc
+extern time_t boot_time;
+
+// JSON compiler definitions.
+class JSONWriter {
+ public:
+  explicit JSONWriter(std::ostream& out)
+      : out_(out), indent_(0), state_(JSONOBJECT) {}
+
+  inline void indent() { indent_ += 2; }
+  inline void deindent() { indent_ -= 2; }
+  inline void advance() {
+    for (int i = 0; i < indent_; i++) out_ << " ";
+  }
+
+  inline void json_start() {
+    if (state_ == JSONVALUE) out_ << ",";
+    out_ << "\n";
+    advance();
+    out_ << "{";
+    indent();
+    state_ = JSONOBJECT;
+  }
+
+  inline void json_end() {
+    out_ << "\n";
+    deindent();
+    advance();
+    out_ << "}";
+    state_ = JSONVALUE;
+  }
+  template <typename T>
+  inline void json_objectstart(T key) {
+    if (state_ == JSONVALUE) out_ << ",";
+    out_ << "\n";
+    advance();
+    out_ << "\"" << key << "\""
+         << ": {";
+    indent();
+    state_ = JSONOBJECT;
+  }
+
+  template <typename T>
+  inline void json_arraystart(T key) {
+    if (state_ == JSONVALUE) out_ << ",";
+    out_ << "\n";
+    advance();
+    out_ << "\"" << key << "\""
+         << ": [";
+    indent();
+    state_ = JSONOBJECT;
+  }
+  inline void json_objectend() {
+    out_ << "\n";
+    deindent();
+    advance();
+    out_ << "}";
+    state_ = JSONVALUE;
+  }
+
+  inline void json_arrayend() {
+    out_ << "\n";
+    deindent();
+    advance();
+    out_ << "]";
+    state_ = JSONVALUE;
+  }
+  template <typename T, typename U>
+  inline void json_keyvalue(T key, U value) {
+    if (state_ == JSONVALUE) out_ << ",";
+    out_ << "\n";
+    advance();
+    out_ << "\"" << key << "\""
+         << ": "
+         << "\"";
+    out_ << value << "\"";
```

Comment:
What happens if either the key or the value need escaping for JSON? I don’t think we currently account for that, do we?

Reply:
keys are not escaped, as we defined the keys, and with no JSON specials.
normal values are not escaped, as I examine those I don't see anything that require escaping.
the only exception is environment variables, which I am escaping (`EscapeJsonChars`).

Reply:
That seems a bit fragile? It’s not true that the only exception is environment variables, there’s also at least CLI args, filenames, and error messages.

I’d prefer to always run `EscapeJsonChars` on this value, and not worry about it at the call site…

Reply:
thanks, now all `values` are JSON escaped. I had to `convert` some of the values / primitive types to c++ strings before processing this, which in part renders the template function `json_keyvalue` and `json_element` ineffective (they always take strings now); but no other side effects.